### PR TITLE
Rest API Conversion to Spring MVC with API Documentation [GEOS-7931]

### PR DESCRIPTION
**Pull request by tbarsballe**
See https://github.com/geoserver/geoserver/wiki/REST-API-Refresh for more details.

* https://osgeo-org.atlassian.net/browse/GEOS-7931
Improved REST API Documentation

### CODE REVIEW

**Comment by jodygarnett** [comment]
Woo hoo!

**Comment by bencaradocdavies** [comment]
Do not merge this pull request. There is still an intermittent failure in gs-h2 ```RestTest```:
```
createDataStoreUsingRestSingleFile(org.geoserver.h2.RestTest)  Time elapsed: 267 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <201>
     but: was <202>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:865)
	at org.junit.Assert.assertThat(Assert.java:832)
	at org.geoserver.h2.RestTest.genericCreateDataStoreUsingRestTest(RestTest.java:123)
	at org.geoserver.h2.RestTest.createDataStoreUsingRestSingleFile(RestTest.java:105)
```
Despite REST URL with ```?configure=all```, sometimes ```source.getNames()``` is empty at line 412 in ```DataStoreFileController.dataStorePut```.

**Comment by bencaradocdavies** [comment]
Also, many modules in the ```communityRelease``` profile do not compile.

**Comment by aaime** [comment]
@bencaradocdavies the idea during the sprint was to kick them out of the build (them being community modules, and thus unsupported). I'd still go on the ml and warn people providing a list of the modules that will be removed.

**Comment by bencaradocdavies** [comment]
@aaime that sounds good. Removal from the community profile should provide incentive to the maintainers to migrate to the new REST implementation. Some of the fixes might be just package name changes, which was all that I had to do for app-schema.

**Comment by jodygarnett** [comment]
Do we have a Jira issue for this? Or many Jira issues ...

**Comment by jodygarnett** [comment]
I am really looking forward to getting this in! Thanks for all the hard work everyone. Also to everyone who worked post-sprint: Andrea (passing tests) and Torben (finishing importer and making the code consistent for this PR.

**Comment by tbarsballe** [comment]
> Do we have a Jira issue for this? Or many Jira issues ...

Good question. As far as I can tell, there is neither a JIRA issue nor a GSIP. This is mostly just documented through wiki pages and mailing list discussion.

Looking back to last years Wicket migration, I don't think we had an JIRA issue for that either.

File: src/restconfig/src/main/java/org/geoserver/rest/catalog/CoverageStoreController.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        CoverageStoreInfo original = catalog.getCoverageStoreByName(workspaceName, storeName);
+        if(original == null) {
+            throw new ResourceNotFoundException(
+                    "No such coverage store: " + workspaceName + "," + storeName);
+        }
+        return original;
+    }
+    
+    @DeleteMapping(value = "{storeName}")
+    public void overageStoreDelete(
```
Comment (jodygarnett) [inline src/restconfig/src/main/java/org/geoserver/rest/catalog/CoverageStoreController.java]: Rename from {{overageStoreDelete}} to {{coverageStoreDelete}}

Comment (jodygarnett) [inline src/restconfig/src/main/java/org/geoserver/rest/catalog/CoverageStoreController.java]: Fixed

**Comment by jodygarnett** [comment]
From conversation with @bencaradocdavies - the H2 failure has occurred on master before:

* http://ares.boundlessgeo.com/jenkins/job/geoserver-master/4475/consoleText

> geoserver-master 4475 is the only one that confirms the failure I saw on this branch (createDataStoreUsingRestSingleFile), but it is confirmed on master, not a new regression, and thus not a blocker for merge.

**Comment by bencaradocdavies** [comment]
Kicked the PR build as geogig is back on master.

File: doc/en/src/main/resources/api/datastores.yaml [hunk 2]
```diff
@@ -0,0 +1,186 @@
+---
```
Comment (tbarsballe) [inline doc/en/src/main/resources/api/datastores.yaml]: Comments in yaml use "#" I believe

**Comment by tbarsballe** [comment]
After the latest CI build, the following community modules are failing:

```
[INFO] Core Scripting Extension ........................... FAILURE [  2.506 s]
[INFO] REST plugin community module ....................... FAILURE [  3.774 s]
[INFO] GeoServer JMS clustering module .................... FAILURE [  5.283 s]
[INFO] Resumable REST Upload module ....................... FAILURE [  1.903 s]
[INFO] GeoFence security integration ...................... FAILURE [  3.659 s]
[INFO] REST SLD service ................................... FAILURE [  2.318 s]
[INFO] GeoGig GeoServer integration ....................... FAILURE [ 19.913 s]
[INFO] GeoServer Config Backup And Restore REST Module .... FAILURE [  1.894 s]
[INFO] GeoServer notification event system ................ FAILURE [  4.166 s]
```

**Pull request merged**